node/subsystem-util/src/database.rs

```diff
@@ -0,0 +1,335 @@
+			DBTransaction::new()
+		}
+
+		fn get(&self, col: u32, key: &[u8]) -> Result<Option<DBValue>> {
+			map_err(self.db.get(col as u8, key))
+		}
+
+		fn get_by_prefix(&self, col: u32, prefix: &[u8]) -> Option<Box<[u8]>> {
+			self.iter_with_prefix(col, prefix).next().map(|(_, v)| v)
+		}
+
+		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+			let mut iter = handle_err(self.db.iter(col as u8));
+			Box::new(std::iter::from_fn(move || {
+				if let Some((key, value)) = handle_err(iter.next()) {
+					Some((key.into_boxed_slice(), value.into_boxed_slice()))
+				} else {
+					None
+				}
+			}))
+		}
+
+		fn iter_with_prefix<'a>(
+			&'a self,
+			col: u32,
+			prefix: &'a [u8],
+		) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+			if prefix.len() == 0 {
+				return self.iter(col)
+			}
+			let mut iter = handle_err(self.db.iter(col as u8));
+			handle_err(iter.seek(prefix));
+			Box::new(std::iter::from_fn(move || {
+				if let Some((key, value)) = handle_err(iter.next()) {
+					key.starts_with(prefix)
+						.then(|| (key.into_boxed_slice(), value.into_boxed_slice()))
+				} else {
+					None
+				}
+			}))
+		}
+
+		fn restore(&self, _new_db: &str) -> Result<()> {
+			unimplemented!("restore is unsupported")
+		}
+
+		fn io_stats(&self, _kind: IoStatsKind) -> IoStats {
+			unimplemented!("io_stats not supported by parity_db");
+		}
+
+		fn has_key(&self, col: u32, key: &[u8]) -> Result<bool> {
+			map_err(self.db.get_size(col as u8, key).map(|r| r.is_some()))
+		}
+
+		fn has_prefix(&self, col: u32, prefix: &[u8]) -> bool {
+			self.get_by_prefix(col, prefix).is_some()
+		}
+
+		fn write(&self, transaction: DBTransaction) -> std::io::Result<()> {
+			let mut ops = transaction.ops.into_iter();
+			// TODO using a key iterator or native delete here would be faster.
+			let mut current_prefix_iter: Option<(parity_db::BTreeIterator, u8, Vec<u8>)> = None;
+			let current_prefix_iter = &mut current_prefix_iter;
+			let transaction = std::iter::from_fn(move || loop {
+				if let Some((prefix_iter, col, prefix)) = current_prefix_iter {
+					if let Some((key, _value)) = handle_err(prefix_iter.next()) {
+						if key.starts_with(prefix) {
+							return Some((*col, key.to_vec(), None))
+						}
+					}
+					*current_prefix_iter = None;
+				}
+				return match ops.next() {
+					None => None,
+					Some(DBOp::Insert { col, key, value }) =>
+						Some((col as u8, key.to_vec(), Some(value))),
+					Some(DBOp::Delete { col, key }) => Some((col as u8, key.to_vec(), None)),
+					Some(DBOp::DeletePrefix { col, prefix }) => {
+						let col = col as u8;
+						let mut iter = handle_err(self.db.iter(col));
+						handle_err(iter.seek(&prefix[..]));
+						*current_prefix_iter = Some((iter, col, prefix.to_vec()));
+						continue
+					},
+				}
+			});
+
+			// Locking is required due to possible racy change of the content of a deleted prefix.
+			let _lock = self.write_lock.lock();
```

Comment:
we might be calling it from an async context and need to be careful not to cause deadlocks